### PR TITLE
Fix: raise error when namelist group is not closed before the start of a new group

### PR DESCRIPTION
**Pull request by dhohn**
Hi, I hope this is okay to submit — I ran into a bug where a missing / terminator on a namelist group would silently cause the following group to be skipped entirely. For example, in a file like:

```
  &open_group
  a = 1

  &closed_group
  b = 2
  /
```

closed_group would be silently dropped from the parsed result, which can be pretty hard to debug. I've looked at the bevahiour of gfortran and ifx for this case. gfortran errors and exits. ifx reads both groups without warning. I prefer an error personally so I made the fix to be an error.

The fix peeks at the next non-whitespace token when & is seen as a group terminator. If it's not a recognised F77-style terminator (&end, standalone &, or $), a ValueError is raised pointing to the offending group. itertools.tee is used to peek without consuming from the token stream, which felt consistent with how lookahead is already handled elsewhere in the parser.

This might also relate to this issue: #160 

Would be happy to adjust anything if this doesn't fit. Thanks for considering it!

### CODE REVIEW

**Comment by marshallward** [comment]
Thanks, this looks very valuable.  As you say, the behavior can vary depending on the compiler, regardless of what is in the language standard.  So it might be a good idea to introduce a flag to control this behavior.

For example, the namelist above could be interpreted as follows.
```
  &open_group   ! <-- start a new group
  a = 1
  &  ! <-- legacy termination token (from &closed_group)
    
  ! Not formally a namelist group, so excluded from data transfer?
  closed_group
  b = 2
  /
  
  ! Continue searching for the next & or EOF.
```
This might still be a formally correct namelist, even some compilers raise an error.  I've never been entirely sure myself.  (I think it depends on what defines a "data transfer".)

I agree that this is a frustratingly subtle error, so it makes sense for this to raise an error on default, with an option to disable it.

I'll try to look over the lookahead a bit later today.  Thanks for putting this together!

File: f90nml/parser.py [hunk 1]
```diff
@@ -384,6 +384,22 @@ def _readstream(self, nml_file, nml_patch_in=None):
...
+                    if self.token == '&':
+                        # Peek at the next non-whitespace token without
+                        # consuming it, to distinguish F77-style terminators
+                        # (&end or standalone &) from an unclosed group.
+                        self.tokens, peek_iter = itertools.tee(self.tokens)
+                        skip = self.comment_tokens + whitespace
+                        next_tok = next(
+                            (t for t in peek_iter if t[0] not in skip), None
+                        )
+                        if next_tok not in ('end', '&', '$', None):
```
[the remark this answers is not in the document]
Comment (marshallward) [reply f90nml/parser.py]: Probably correct, I don't think case-insensitivity has been applied yet.  `next_tok.lower()` ?

File: f90nml/parser.py [hunk 2]
```diff
@@ -384,6 +384,22 @@ def _readstream(self, nml_file, nml_patch_in=None):
                 # Finalise namelist group
                 if self.token in ('/', '&', '$'):
 
+                    if self.token == '&':
```
Comment (marshallward) [inline f90nml/parser.py]: Should this test for both '&' and '$' ?

For what it's worth, NVIDIA appears to happily match `&grp` with `$end` and every combination therein.

Comment (dhohn) [reply f90nml/parser.py]: Yes. Oh I didnt even consider mixed opening and closing tokens. That would become really tricky with multiple groups with different conventions in one file. I would hope such files just dont exists (but you never know with fortran ;)

File: f90nml/parser.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhohn) [reply f90nml/parser.py]: Yes. Ive also added a test.

**Comment by dhohn** [comment]
Hi @marshallward ,

Thanks for having a look and commenting. I have taken up both your suggestions above and added tests for those cases as well.

Regarding the mixed open/close token usage, thats probably an edge case of an edge case if at all. Do you want to do anything about that? The current code also allows any combination of tokens to open and close.

**Comment by marshallward** [comment]
Sorry for losing track of this one.  Any comments that I had were very minor and can be adjusted at a later time.  Thanks again for putting this together.
